tests/metrics/test_auc.py

```diff
@@ -0,0 +1,285 @@
+    )
+    def test_when_no_interactions(self, insufficient_handling: str) -> None:
+        reco = pd.DataFrame([[1, 1, 1], [2, 1, 1]], columns=[Columns.User, Columns.Item, Columns.Rank])
+        expected_metric_per_user = pd.Series(index=pd.Series(name=Columns.User, dtype=int), dtype=np.float64)
+        metric = PAP(k=3, insufficient_handling=insufficient_handling)
+        pd.testing.assert_series_equal(metric.calc_per_user(reco, EMPTY_INTERACTIONS), expected_metric_per_user)
+        assert np.isnan(metric.calc(reco, EMPTY_INTERACTIONS))
+
+    @pytest.mark.parametrize("k", (1, 3))
+    @pytest.mark.parametrize(
+        "insufficient_handling", (InsufficientHandling.RAISE, InsufficientHandling.EXCLUDE, InsufficientHandling.SKIP)
+    )
+    def test_when_duplicates_in_interactions_sufficient(self, k: int, insufficient_handling: str) -> None:
+        reco = pd.DataFrame(
+            {
+                Columns.User: [1, 1, 1, 2, 2, 2],
+                Columns.Item: [1, 2, 3, 1, 2, 3],
+                Columns.Rank: [1, 2, 3, 4, 5, 6],
+            }
+        )
+        interactions = pd.DataFrame(
+            {
+                Columns.User: [1, 1, 1, 2, 2, 2],
+                Columns.Item: [1, 2, 1, 1, 2, 3],
+            }
+        )
+        metric = PAP(k=k, insufficient_handling=insufficient_handling)
+        expected_metric_per_user = pd.Series(
+            [1, 0],
+            index=pd.Series([1, 2], name=Columns.User),
+            dtype=float,
+        )
+        pd.testing.assert_series_equal(metric.calc_per_user(reco, interactions), expected_metric_per_user)
+
+    # this one is actually useful
+    @pytest.mark.parametrize(
+        "k, insufficient_handling, expected_pauc, expected_users",
+        (
+            (2, InsufficientHandling.SKIP, [1 / 2, 0], [1, 2]),
+            (2, InsufficientHandling.EXCLUDE, [0], [2]),
+        ),
+    )
+    def test_when_duplicates_in_interactions_insufficient(
+        self, k: int, insufficient_handling: str, expected_pauc: tp.List[int], expected_users: tp.List[int]
+    ) -> None:
+        reco = pd.DataFrame(
+            {
+                Columns.User: [1, 2, 2, 2],
+                Columns.Item: [1, 1, 2, 3],
+                Columns.Rank: [1, 4, 5, 6],
+            }
+        )
+        interactions = pd.DataFrame(
+            {
+                Columns.User: [1, 1, 1, 2, 2, 2, 1],
+                Columns.Item: [1, 2, 1, 1, 2, 3, 10],  # last positive is not in reco
+            }
+        )
+        metric = PAP(k=k, insufficient_handling=insufficient_handling)
+        expected_metric_per_user = pd.Series(
+            expected_pauc,
+            index=pd.Series(expected_users, name=Columns.User),
+            dtype=float,
+        )
+        pd.testing.assert_series_equal(metric.calc_per_user(reco, interactions), expected_metric_per_user)
+
+
+# # # TODO: exclude -> nan?
```

Comment:
Haven't decided if we need to return None for excluded users or just no return them at all in 'calc_per_user' methods.

Reply:
Let's exclude them. It's faster and safer. User can easily fill them with nans using reindex